### PR TITLE
SearchableSnapshotIndexInput should read all bytes

### PR DESCRIPTION
**Pull request by tlrx**
This pull request fixes a bug in `SearchableSnapshotIndexInput` which does not use the `InputStream.read(byte[], int, int)` method correctly.

The javadoc indicates that:
> An attempt is made to read as many as <code>len</code> bytes, but a smaller number may be read. 

Consequently, `SearchableSnapshotIndexInput` should try to read more bytes when a first reading operation has not brought back all the requested bytes.

### CODE REVIEW

**Comment by DaveCTurner** [comment]
D'oh, of course. Change looks good, but can we engineer a test that fails because of this bug too?

**Comment by tlrx** [comment]
@DaveCTurner the bug was here since the beginning and passed all reviews. I'm just happy that adding S3 integration tests caught this.

> but can we engineer a test that fails because of this bug too?

Sure, I should have added one, just doing ten things at a time :sweat: 

I pushed https://github.com/elastic/elasticsearch/pull/52199/commits/4de4f81846150a0f1edf6174b55823aaac5f6f0a which randomized how read bytes operations are served. I did not add a specific test but we already have assertions, intensive randomized read tests and we'll soon have integration tests so I think we're good. Unless you think otherwise?

File: x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/SearchableSnapshotIndexInput.java [hunk 1]
```diff
@@ -121,9 +121,19 @@ private void readInternalBytes(final int part, long pos, final byte[] b, int off
...
                      = blobContainer.readBlob(fileInfo.partName(part), pos + optimizedReadSize, length - optimizedReadSize)) {
-                final int directReadSize = inputStream.read(b, offset + optimizedReadSize, length - optimizedReadSize);
+
+                int directReadSize = 0;
+                while ((optimizedReadSize + directReadSize) < length) {
+                    final int read = inputStream.read(b,
+                                                      offset + optimizedReadSize + directReadSize,
+                                                      length - optimizedReadSize - directReadSize);
+                    if (read == -1) {
+                        break;
```
Comment (DaveCTurner) [inline x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/SearchableSnapshotIndexInput.java]: This means we hit the end of the file, which I think is unexpected. In tests the assertion below will fail; in production we will proceed even though some of the buffer contains its old contents, which would be bad. I think instead we should throw an `EOFException` here.

File: x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/SearchableSnapshotIndexInput.java [hunk 2]
```diff
@@ -276,10 +286,17 @@ boolean canContinueSequentialRead(int part, long pos) {
 
         int read(byte[] b, int offset, int length) throws IOException {
             assert this.pos < maxPos : "should not try and read from a fully-read stream";
-            int read = inputStream.read(b, offset, length);
-            assert read <= length : read + " vs " + length;
-            pos += read;
-            return read;
+            int totalRead = 0;
```
Comment (DaveCTurner) [inline x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/SearchableSnapshotIndexInput.java]: Could we extract a method to do this? The two changes to this file look very similar.

**Review by DaveCTurner — DISMISSED** [review]
I left a couple more comments.

File: x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/SearchableSnapshotIndexInput.java (hunk 1 above)
Comment (tlrx) [reply x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/SearchableSnapshotIndexInput.java]: > I think instead we should throw an `EOFException` here.

I pushed e06a94b0f558ed2b85fdf1975856141a1d42210e

File: x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/SearchableSnapshotIndexInput.java (hunk 2 above)
Comment (tlrx) [reply x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/SearchableSnapshotIndexInput.java]: Sure, done in e06a94b0f558ed2b85fdf1975856141a1d42210e

**Comment by tlrx** [comment]
@DaveCTurner Thanks for your review. I've applied your feedback, would you mind to have another look?

File: x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/SearchableSnapshotIndexInput.java [hunk 3]
```diff
@@ -276,7 +296,7 @@ boolean canContinueSequentialRead(int part, long pos) {
 
         int read(byte[] b, int offset, int length) throws IOException {
             assert this.pos < maxPos : "should not try and read from a fully-read stream";
-            int read = inputStream.read(b, offset, length);
+            final int read = readFully(inputStream, b, offset, length, () -> {});
```
Comment (DaveCTurner) [inline x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/SearchableSnapshotIndexInput.java]: I'm wondering whether we should throw an EOFException if this hits EOF too. Do we need to behave differently here?

Comment (DaveCTurner) [reply x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/SearchableSnapshotIndexInput.java]: Oh no I get it now, we might use this to read past the end of a stream that we previously opened. Never mind.

**Review by DaveCTurner — APPROVED** [review]
LGTM

File: x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/SearchableSnapshotIndexInput.java (hunk 3 above)
Comment (tlrx) [reply x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/SearchableSnapshotIndexInput.java]: Yes

**Comment by tlrx** [comment]
Waiting for #52258 to be merged first.

**Comment by tlrx** [comment]
Thanks David

**Pull request merged**